### ISSUE TITLE
Scaling of images for toolbaricons and imagecells

### ISSUE THREAD

**Issue by TomQv** [picoe/Eto#1312]
What is the proper way to correctly prepare images for toolbaricons and imagecells for sharp and clear graphics?

I have all my graphics as svg so I'm able to generate best quality pngs at all sizes.
Currently I create a png at a specific size from the svg and use this as toolbaricon.
If I make the png bigger, I get bigger toolbarbuttons.

At some point the screen-dpi must come in, but where in eto?

For example if I change Windows screen-settings to 150% my icons just become bigger but still unsharp.
Only with Windows-screen scaling 100% = 96dpi everything is perfect.

**Comment by TomQv** [comment]
Pls help what eto.wpf is doing with tool bar-icons on higher screen-resolutions.
Can't figure it out, but all icons are scaled up which looks very blurry.

**Comment by TomQv** [comment]
Ok, I figured out, that we can control this with a app.manifest.
see here: https://stackoverflow.com/questions/23551112/how-can-i-set-the-dpiaware-property-in-a-windows-application-manifest-to-per-mo/44009779

Adding this block to the app.manifest prevents windows from scaling images and its now up to me to scale everything properly, what I actually wanted:

```
<asmv3:application>
  <asmv3:windowsSettings>
    <dpiAware xmlns="http://schemas.microsoft.com/SMI/2005/WindowsSettings">true/pm</dpiAware> <!-- legacy -->
    <dpiAwareness xmlns="http://schemas.microsoft.com/SMI/2016/WindowsSettings">permonitorv2,permonitor</dpiAwareness> <!-- falls back to pm if pmv2 is not available -->
  </asmv3:windowsSettings>
</asmv3:application>
```
Now I only need to figure out, how to scale ui-fonts

**Comment by cwensley** [comment]
Is this still an issue?  Adding the app.manifest is a way to do that (if you aren't targeting a newer version of .NET like 4.7).  Let me know, thanks!

**Comment by TomQv** [comment]
Yes, its still an issue and even a big one.

Adding the manifest works, the app is not scaled any more by Windows so I can use my nice svg, but then all other controls and all fonts are too small.

So disabling Windows scaling is not a good option.

I need a way to use svg as toolbar-buttons so that they are not scaled by windows anymore on high-res screens. But all other controls must still be scaled.

**Comment by cwensley** [comment]
@TomQv I'm not sure I am seeing the same issue as you then, do you have a sample?  The Eto.Test.Wpf application uses [this app.manifest](https://github.com/picoe/Eto/blob/develop/test/Eto.Test.Wpf/app.manifest) and appears to work just fine with the toolbar and image cells.

Also, I'm not sure how you are utilizing .svg files, it isn't a supported option afaik.  Perhaps it's in the way you are converting them to a bitmap (assuming you are doing so).

**Comment by TomQv** [comment]
Right, svg is not supported natively. So on start of the app I check for the screen-dpi and render the toolbar icons into pngs with the respective size (using SkiaSharp.Svg). For example, I want them in 20x20, screen is 150% ´, so I create 30x30 png (perfectly sharp!) which is then used for the toolbarbutton-image.

Unfortunate, Windows is then also scaling them up by 150% and then they become blurry.

Dont know, how to tell Windown, not to touch my pngs, but still scale all other control.

**Comment by cwensley** [comment]
Try `Bitmap.WithSize(logicalWidth, logicalHeight)`, after you have it in a bitmap object.

Otherwise, yes, it will scale the image up as it will think it is not a high dpi image.

**Comment by cwensley** [comment]
.. or more precisely..

```c#
myToolItem.Image = myBitmap.WithSize(20, 20);
```

**Comment by TomQv** [comment]
doesn't work, I mean, still scaled up afterwards.

The trick might be the dpi-tag in the png-file. 
I noticed already that this is affecting for example gridItems.
All my griditem images are 16x16 pngs. But they all must also have the same dpi, otherwise the grid is showing up with different gridrow heights.

So my idea is, I just need the svg rendered into a png with screen-resolution-dpi-
Must figure out, how to set the png-tag to, before I assign to the toolbarbutton-image. 
Skia doesn't support this.

**Comment by TomQv** [comment]
Maybe the problem is Eto.
I was just trying to understand, whats actually happen step by step, specially on the Wpf-side, when I just do:
`var button = new Button {Image = myImage };`
Sorry, I gave up, so complicated, impossible to understand. On the other hand, there are lots of options regarding bitmaps and dpi in wpf, maybe Eto is not using them correctly.

**Comment by cwensley** [comment]
@TomQv, hm sorry it's so complicated.  Not all platforms read the DPI from the png/bitmap, so using  `Eto.Drawing.Icon` is the recommended way to support that, which the `WithSize` method returns.

With my testing it works well with Wpf, so if you have a sample of what you're doing that isn't working for you I can take a look to see what may be going wrong.

Thanks!

**Comment by TomQv** [comment]
In Eto.Test.Wpf you just provide very big toolbar-images.
Like this you are on the safe side, because they always being scaled down, never up, whatever dpi the screen is.  But still they are scaled, losing quality

I don't want this, I want the exact, unscaled image as toolbar image.
But I also don't have more time currently for this.

**Comment by TomQv** [comment]
Ok, I think, I figured out, why my code is not working with dpi and scalings, but Eto.Test is.
Apart from using the app.manifest from Eto.test this line in app.config is also required:
Don't know why and I also don't know, how one is supposed to know this, but including this line, everything works.
  ```
 <!-- Enable .NET 4.6.2 fixes for high-dpi even though we are built for .NET 4.5 -->
    <AppContextSwitchOverrides value="Switch.MS.Internal.DoNotApplyLayoutRoundingToMarginsAndBorderThickness=false;Switch.System.Windows.DoNotScaleForDpiChanges=false"/>
```

**Comment by cwensley** [comment]
@TomQv, glad you figured it out.  Yes, that line is necessary if you are compiling your .exe with .NET Framework < 4.6.2.  If you compile with >= 4.6.2 then that line won't be needed afaik.

Did that fully resolve this issue for you is there anything outstanding?

**Comment by TomQv** [comment]
Its a workaround, and I don't like it, but its fine, at least regarding Eto.
Thanks
Tom

**Comment by cwensley** [comment]
> Its a workaround

How is it a workaround?  This is how you get WPF to be fully DPI aware, regardless of Eto.  Or are you talking about something else being a workaround?

**Comment by TomQv** [comment]
I did it now like you in Eto.Test, just providing my Icons in 100x100 pixels to make sure they will never be upscaled. But WPF will always downscale then, so we always have a slight loss in quality.
This is why I took all the effort to implement svg to have the icons created at runtime exactly in the size we need. But that was useless, wasting time.

**Comment by cwensley** [comment]
@TomQv ah, I got confused by your comment:

>but including this line, everything works.

It sounds like this is still not working in that you can't get it to render pixel perfect images.

**Comment by TomQv** [comment]
No, its not pixel perfect, it cant be and I gave up.
But the quality is now acceptable, so its fine.

**Comment by TomQv** [comment]
Ok, here is the final results after more investigations:

Button-images are pixel-perfect when you set Button.Size =new Size(-1,-1) and **Windows screen-scale is 100%**.  Only then!
So when I want by Button-images 20x20, I render the svg into a png 20x20 and the button is pixel-perfect.

If Windows screen-scale is larger then 100%, then wpf is scaling the images by this factor. 
Thats not in eto and there is nothing we can do.

For example Windows screen-scale 150%.
So I render my png as 30x30 and I end up with a 45x45 button, obviously blurred, because wpf is still upscaling it by factor 1.5.

Workaround is:
I set Button.Size =new Size(30,30) and render svg to 30x30-
Image is still scaled so not pixel-perfect, but at least not so bad.

Why am I doing all this?
Our app is also used in automotive use cases, in cars for example, so the user can choose the size of the buttons in a wide range to best fit his or her need while driving.

Maybe if I find some time I will try the following:
The Eto.Bitmap->ToWpf conversion always creates a wpf-imagesource with 96dpi.
Would be interesting, if we could set this to the actual screen-dpi, maybe it wouldn't be scaled then.

Anyway, all good now, you can close this issue.

**Comment by LaraSQP** [comment]
How about icon fonts?
